### ISSUE TITLE
Improvements on code generation

### ISSUE THREAD

**Issue by paulo-ferraz-oliveira** [tomas-abrahamsson/gpb#188]
Hi.

I was `dialyzer`'ing `hex_core` and [came into two potential issues](https://github.com/hexpm/hex_core/pull/105#issuecomment-718255421) linked to `gpb`'s generated code.

---

```erlang
src/hex_pb_package.erl
1395: The pattern <X, Path> can never match since previous clauses completely covered the type <integer(),['Package' | 'Release' | 'RetirementStatus' | 'reason' | 'releases' | 'retired',...]>
```

is related to the fact that (extra - unused) code is being generated for an internal function. The `dialyzer` infers that the code is only called for integers so assumes the extra "catch-all/no-integer" branch of execution shouldn't exist.

---

```erlang
src/hex_registry.erl
  83: Function sign_protobuf/2 has no local return
  85: The call hex_pb_signed:encode_msg(#{'payload':=binary(), 'signature':=binary()}, 'Signed') will never return since the success typing is (#{'payload':=binary() | maybe_improper_list(binary() | maybe_improper_list(any(),binary() | []) | byte(),binary() | [])}, 'Signed') -> binary() and the contract is ('Signed'(), atom()) -> binary()
```

is related to the fact that an optional field (in a message) isn't showing up as an optional key in the generated code, which leads `dialyzer` to believe it can't be used.

---

Are you aware of these issues? Would you be open to a fix, considering they're important to you?

Many thanks.

---

**Note**: `hex_core` uses `gpb` via `rebar3_gpb_plugin` which I updated to latest to see if the issues would go away but they didn't. I haven't tried `hex_core` with the latest `gpb` directly.

**Comment by tomas-abrahamsson** [comment]
The intention is that generated code should be free from dialyzer warnings. I noticed the vendored hex_pb_package was from gpb 4.2.1. Have you tried with a newer gpb? (Not at the computer at the moment, so please forgive if I'm missing something)

**Comment by paulo-ferraz-oliveira** [comment]
> Have you tried with a newer gpb?

I haven't but I will now. I like that the "intention is that generated code should be free from dialyzer warnings".

**Comment by tomas-abrahamsson** [comment]
Gpb can generate diffeeent kinds of code (records, maps of a few variants) and there can be many different kinds of proto definitions and dialyzer has been a bit too slow to run in all unit tests, so it has traditionally been a best effort intention. At times, I've fixed the issues I've seen.

**Comment by paulo-ferraz-oliveira** [comment]
> it has traditionally been a best effort intention

Not judging 😄. Have always used `gpb` without issues (this is the first one I've noticed, as far as `dialyzer`'ing goes), and been doing so for years now. I don't know if I'm using the same options to generate the `.erl` as `hex_core` is though, but I suspect it wouldn't make a difference in these two cases.

I'm willing to give the fix a shot, but I haven't yet looked into the code.

**Comment by tomas-abrahamsson** [comment]
No problems, and you're most welcome to take give it a shot!

**Comment by tomas-abrahamsson** [comment]
I managed to reproduce. I noticed the `{target_erlang_version, 17}` in hex_core's rebar.config. The issue in hex_pb_package.erl goes away if generating for Erlang 18 or later, since gpb then inserts `-dialyzer({nowarn_function,v_type_sint32/3}).` for the verifier auxiliary function, but I don't think that was possible in Erlang 17, so I'm unsure if this is easy or difficult to fix. (I'm fearing it could mean some bigger restructurings of the generated code are needed)

The other issue, the one in hex_pb_signed.erl goes away if generating for Erlang 19 or later. I think the issue here is that there was a change in type spec semantics when Erlang 19 arrived. In Erlang 17 and 18, additional fields were implicitly allowed to the map even if not in the map type, but with the Erlang 19 one had to explicitly say that, for exampe with `_ => _` or similar. (That's why gpb generates optional fields as out-commented when for Erlang 17 or 18) So I think if one would use a dialyzer from Erlang 17 or 18, this issue might probably go away, but if using dialyzer >= 19 on this code intended for Erlang 17, the issue is because of changed interpretation of type specs for maps. 

(Side note: I think gpb is doing something wrong here. It indicates the required field with `=>` and the optional fields as out-commented. For Erlang 17 and 18, it should probably indicate _both_ fields as out-commented. In Erlang 19 and later, type specs have `:=` to indicate mandatory map associations and `=>` for map assocs that may exist. I first used these to indicate `required` or `optional` but since then I've come to realize that gpb must always declare fields as `=>` (optional if you will), since on decoding, gpb cannot guarantee presence; if the input binary-to-decode does not contain the field, gpb will set it, hence even required fields are optional from a strict point of view.)

**Comment by paulo-ferraz-oliveira** [comment]
> I'm unsure if this is easy or difficult to fix

**Comment by paulo-ferraz-oliveira** [comment]
> I'm unsure if this is easy or difficult to fix

You mean "if we decide to fix `gpb`", right? I can propose `hex_core` to move up to 18 (or even 19).

**Comment by paulo-ferraz-oliveira** [comment]
> For Erlang 17 and 18, it should probably indicate both fields as out-commented

Agreed.

> In Erlang 19 and later, type specs have := to indicate mandatory map associations and => for map assocs that may exist.

Agreed.

This would be the only fix required to `gpb` in any case, it seems, right?

**Comment by tomas-abrahamsson** [comment]
> You mean "if we decide to fix `gpb`", right? I can propose `hex_core` to move up to 18 (or even 19).

True, I would much prefer that. Given the age of Erlang 17 and the effort probably needed, fixing it doesn't look too tempting :) Preferably 19 if possible in your case, since then I think you will be able to run dialyzer from any 19-or-later Erlang[*]. If you would go for 18 (prior to the semantic change to map type specs), I think you would also need to use dialyzer from Erlang 18, or maybe you do that already?

__
[*] at the moment, at least. Erlang might of course change type specs again in the future, in which case this wouldn't hold. (Type specs could change for new types or new syntax, that might require a few major version or so to settle) 

> This would be the only fix required to gpb in any case, it seems, right?

(assuming you referring to "For Erlang 17 and 18, it should probably indicate both fields as out-commented") Yes, agreed, this ought to happen. Although in your use case, this will neither add nor reduce the number of dialyzer issues; running the Erlang-17 dialyzer would (as long as 17 is targeted.) I'll make a note of fixing it eventually, unless you beat me to it of course. I think it has lower prio, since I think Erlang 17 or 18 users is not a big group, and it is already fixed for 19 or later.

**Comment by paulo-ferraz-oliveira** [comment]
> True, I would much prefer that

It's not possible, [according to `hex_core`'s maintainers](https://github.com/hexpm/hex_core/pull/105#issuecomment-720137496). Yeah, I don't think "fixing" `gpb` in this sense (`dialyzer` stuff) would benefit it all that much, especially if it's complex and/or time-consuming.

> maybe you do that already

I use OTP 19+ only.

> Erlang might of course change type specs again in the future

True.

> this will neither add nor reduce the number of dialyzer issues

If there's anybody else complaining about it, it might be worth giving it a shot; in any case, I don't use OTP 17 or 18, though...

I'm going to try and figure out if there's alternatives to reducing warnings to 0, in `hex_core` and eventually come back to this, but I foresee `gpb` will not suffer changes and this issue'll be simply closed.

**Comment by tomas-abrahamsson** [comment]
About the hex_pb_package.erl issue "The pattern <X, Path> can never match ...": I pushed a fix to the branch fix-erl17-enum-verifier-dialyzer.

On that branch, with regenerated pb files, hex_core dialyzes sort-of-cleanly for me:

```
$ dialyzer --plt x-18.3.4.9.plt -r ebin                                                                                                                                          
  Checking whether the PLT x-18.3.4.9.plt is up-to-date... yes
  Proceeding with analysis...
leexinc.hrl:268: Function yyrev/2 will never be called
Unknown functions:
  ...
 done in 0m3.56s
done (warnings were emitted)
```
I had no Erlang 17 around with a working dialyzer, but Erlang 18 too is before the map type spec change.

(versions involved in the above: gpb = 5e10b49 and hex-core = 37d336c )

**Comment by paulo-ferraz-oliveira** [comment]
@tomas-abrahamsson, thanks for working on this. That's great news.

If you have `rebar3` installed, it should suffice to use Erlang/OTP 17 and `rebar3 dialyzer` (the PLT should be generated for you - the first time it takes a bit longer).

**Comment by paulo-ferraz-oliveira** [comment]
I think I may have found a solution to this issue, that depends on a vendoring variable (which we can use to target a non-default `=17` Erlang/OTP version).

Do you think there's anymore this Issue should handle (or, e.g., you want to use it to reference a PR)? Or is it ready to get closed?

**Comment by tomas-abrahamsson** [comment]
Thanks for the update, I'm in the process of fixing the other issue with fixing type specs for required fields in Erlang 17, but the code needed some refactoring first. I can close when I merge that in a little while.

**Comment by tomas-abrahamsson** [comment]
Both fixes now included in 4.15.2, (ie: the dialyzer issue in the verifiers, and the out-commentation of fields for maps when generating for Erlang 17 or 18), so I'll be closing this. Thanks for reporting, and please feel free to re-open or open another issue if there's anything more.
